Propagate location from destination alloca in `salvageDebugInfo()`.

asl opened the pull request:
`salvageDebugInfo` is called during SIL Mem2Reg and could produce misleading debug info in the following case:

```
%a = alloc_stack $MyModel.TangentVector, var, name "self", argno 1, implicit, loc "debug2.swift":37:17 ...
...
store %b to %a : $*MyModel.TangentVector
```

Such SIL could be created as a result of inlining (where store comes from the inlined function).
Before this patch we'd end with `debug_value` instruction with variable information, but without location.

This caused LLVM IR debug info verifier assertions when there might be another instruction with complete debug info (including location) for the same argument.

After this patch, when store itself does not contain a valid location information, reuse it from destination alloca.

Also adding implementation of `NonSingleValueInstruction::dump()` while here (e.g. for `DebugValueInst` dumping / printing).

Fixes #58660

asl committed:
Propagate location from destination alloca in `salvageDebugInfo()`.

`salvageDebugInfo` is called during SIL Mem2Reg and could produce
misleading debug info in the following case:

```
%a = alloc_stack $MyModel.TangentVector, var, name "self", argno 1, implicit, loc "debug2.swift":37:17 ...
...
store %b to %a : $*MyModel.TangentVector
```

Such SIL could be created as a result of inlining (where store comes from the inlined function).
Before this patch we'd end with `debug_value` instruction with variable information, but
without or incorrect location.

This caused LLVM IR debug info verifier assertions when there might be another instruction with
complete debug info (including location) for the same argument.

After this patch we always reuse it from destination alloca

Fixes #58660

asl commented:
Tagging @BradLarson

asl commented:
@swift-ci Please test

adrian-prantl reviewed (commented):
Are there any cases where we would not want to take the source location of the AllocStackInst? If not it would be more straightforward to unconditionally take it.

asl commented on lib/SILOptimizer/Utils/InstOptUtils.cpp:
```diff
@@ -1800,15 +1800,22 @@ void swift::salvageDebugInfo(SILInstruction *I) {
+          bool IsValidLoc =
+            (Loc.isFilenameAndLocation() ?
+             Loc.getFilenameAndLocation()->line != 0 :
+             Loc.getSourceLoc().isValid());
```
@adrian-prantl We're having the following:

```
SourceLoc SILLocation::getSourceLoc() const {
  if (isSILFile())
    return storage.sourceLoc;

  // Don't crash if the location is a FilenameAndLocation.
  // TODO: this is a workaround until rdar://problem/25225083 is implemented.
  if (getStorageKind() == FilenameAndLocationKind)
    return SourceLoc();

  return getSourceLoc(getPrimaryASTNode());
}
```

So in case of `isFilenameAndLocation() == true` we'd always end with empty source loc. I'm not 100% sure this is a real case here though :)

asl commented:
After thinking more, I believe we need to always propagate location from alloca. And here is the rationale: consider there is source code location on the store which is not coincide with the source code location on alloca. Since we are propagating variable information, we'll end with something like this:

```
%a = alloc_stack $MyModel.TangentVector, var, name "self", argno 1, implicit, loc "debug2.swift":37:17 ...
...
debug_value %b : $*MyModel.TangentVector. var, name "self", argno 1, implicit, loc "debug2.swift":42:123
```

which is certainly wrong. 

@adrian-prantl Am I missing something?

asl commented:
@adrian-prantl Yes, apparently it's only 1 letter change in the end ;)

asl commented:
@swift-ci please test

asl committed:
Refine logic for source location propagation:
  - Always use alloca_inst source location for arguments as this must be unique
  - Prefer alloca_inst source location when store's location is artificial

asl commented:
> I think I agree. That would also simplify the patch.

@adrian-prantl And I was wrong :) The answer is SROA. It turns alloc_stack like this:

```
  %4 = alloc_stack $MyStruct, var, name "my_struct", loc "sroa.swift":8:9
```
into:
```
  %4 = alloc_stack $Int64, var, (name "my_struct", loc "sroa.swift":8:9, scope 2), type $*MyStruct, expr op_fragment:#MyStruct.x, loc "<compiler-generated>":0:0, scope 2 // users: %20, %17, %15
  %5 = alloc_stack $Int64, var, (name "my_struct", loc "sroa.swift":8:9, scope 2), type $*MyStruct, expr op_fragment:#MyStruct.y, loc "<compiler-generated>":0:0, scope 2 // users: %19, %18, %16
```

This logic is due to `SILDebugVariable::createFromAllocation` which moves source code location into `VarInfo` and create artificial "outer" source code loc. So, it seems we'd need to end with more complex logic. I'm leaning towards always propagating alloca source loc for arguments as these locations must be unique as well as when store source location is artificial

asl commented:
@swift-ci  Please test

asl commented:
@swift-ci please test

asl committed:
Try to unbreak test on Linux

asl commented:
@swift-ci please test
